templates/commands/cel.go

```diff
@@ -0,0 +1,132 @@
+	"time"
+
+	"github.com/abcxyz/abc/templates/model"
+	"github.com/abcxyz/pkg/logging"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+var (
+	celRegistry = types.NewEmptyRegistry()
+
+	celFuncs = []cel.EnvOption{
+		// Adds a split method on strings. Example: "foo,bar".split(",")
+		//
+		// Design decision: use a method instead of a plain function, because
+		// that is the convention that the community seems to prefer for string
+		// splitting. This is based on some quick googling of other peoples'
+		// custom split functions.
+		cel.Function(
+			"split",
+			cel.MemberOverload(
+				"string_split",
+				[]*cel.Type{cel.StringType, cel.StringType},
+				types.NewListType(types.StringType),
+				cel.BinaryBinding(func(lhs, rhs ref.Val) ref.Val {
+					toSplit := lhs.Value().(string)
+					splitOn := rhs.Value().(string)
+					tokens := strings.Split(toSplit, splitOn)
+					return types.NewStringList(celRegistry, tokens)
+				}),
+			),
+		),
+	}
+)
+
+// celCompile parses and compiles the given expr into executable Program.
+func celCompile(ctx context.Context, scope *scope, expr model.String) (cel.Program, error) {
+	startedAt := time.Now()
+
+	celOpts := []cel.EnvOption{}
+	for varName := range scope.All() {
+		celOpts = append(celOpts, cel.Variable(varName, cel.StringType))
+	}
+	celOpts = append(celOpts, celFuncs...) // Add custom function bindings
+
+	env, err := cel.NewEnv(celOpts...)
+	if err != nil {
+		return nil, expr.Pos.AnnotateErr(fmt.Errorf("internal error: failed configuring CEL environment: %w", err))
+	}
+
+	ast, issues := env.Compile(expr.Val)
+	if err := issues.Err(); err != nil {
+		return nil, expr.Pos.AnnotateErr(fmt.Errorf("failed compiling CEL expression: %w", err))
+	}
+
+	prog, err := env.Program(ast)
+	if err != nil {
+		return nil, expr.Pos.AnnotateErr(fmt.Errorf("failed constructing CEL program: %w", err))
+	}
+
+	logger := logging.FromContext(ctx).Named("celCompile")
+	latency := time.Since(startedAt)
+	logger.Debugw("CEL compilation and loading time", "duration_usec", latency.Microseconds())
+
+	return prog, nil
+}
+
+// celEval runs a previously-compiled CEL Program (which you can get from
+// celCompile()).
+//
+// The output of CEL execution is written into the location pointed to by
+// outPtr. It must be a pointer. If the output of the CEL expression can't be
+// converted to the given type, then an error will be returned. For example, if
+// the CEL expression is "hello" and outPtr points to an int, an error will
+// returned because CEL cannot treat "hello" as an integer.
+func celEval(ctx context.Context, scope *scope, pos *model.ConfigPos, prog cel.Program, outPtr any) error {
+	startedAt := time.Now()
+
+	// The CEL engine needs variable values as a map[string]any, but we have a
+	// map[string]string, so convert.
+	scopeMapAny := make(map[string]any, len(scope.All()))
+	for varName, varVal := range scope.All() {
```

Comment:
Nit: can `scope.All()` change between invocations or is it an expensive calculation? If so, we might want to grab it as a local variable once.

Reply:
I think the code as written here only invokes it once. Maybe I'm missing your point though.

Reply:
It's called on L103 and L104. I'm curious if the result is (1) expensive or (2) can change between invocations. Because then we might want:

```go
	scopes := scope.All()
	scopeMapAny := make(map[string]any, scopes)
	for varName, varVal := range scopes {
```

Reply:
Oh of course, fixed. Thanks for catching that.